### PR TITLE
invoices+htlcswitch: fix spurious htlc rejections from overly strict onion packet integrity checks

### PR DESCRIPTION
**Pull request by ProofOfKeags**
## Change Description
This change addresses #7589. The spec was updated [here](https://github.com/lightning/bolts/pull/1032/files) and [here](https://github.com/lightning/bolts/pull/1063) to relax the final hop htlc integrity checks, allowing the final htlc to have a higher payment amount or a higher cltv delta than what is specified in the onion packet.

**Please take special care to review the test cases**. I believe they currently capture all of the invariants we are trying to satisfy as specified in the spec change. However, it appears as though there is some ambiguity related to what these checks are for. According to the spec these are integrity checks to make sure that the incoming HTLC is *more favorable* than what is specified in the packet (presumably to detect any tampering along the route). However, what is unclear here is whether or not the amount/expiry specified in the onion payload is what we should use to compare to invoice parameters, or whether just checking the inbound HTLC parameters against our invoice is sufficient.

## Steps to Test
Steps for reviewers to follow to test the change.

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

- - -
This change is [Reviewable](https://reviewable.io/reviews/lightningnetwork/lnd/7768)

### CODE REVIEW

File: invoices/update.go [hunk 1]
```diff
@@ -214,11 +214,6 @@ func updateMpp(ctx *invoiceUpdateCtx, inv *Invoice) (*InvoiceUpdateDesc,
 	// Add amount of new htlc.
 	newSetTotal += ctx.amtPaid
 
-	// Make sure the communicated set total isn't overpaid.
-	if newSetTotal > ctx.mpp.TotalMsat() {
```
Comment (positiveblue) [inline invoices/update.go:218]: If we delete this this check we need to change the check for "when this set is complete" from
```
setComplete := newSetTotal == ctx.mpp.TotalMsat() // line 241
```

to

```
setComplete := newSetTotal >= ctx.mpp.TotalMsat()
```

**Review by positiveblue — COMMENTED** [review]
Thank you very much for taking care of this issie @ProofOfKeags :tada: 

I did a first pass and I think that we need another small change to  settle invoices that are overpaid.

I also wonder if settling an invoice before receiving all the htlcs can be problematic from the payer PoV:

- Ex: I split a payment in 4, the three first htlcs arrive to you and overpay the invoice, you release the preimage for the first three and the forth is never settled.

I am not saying it is a problem, the logic that we have may already handle this, but just giving it a though.

File: invoices/update.go (hunk 1 above)
Comment (ProofOfKeags) [reply invoices/update.go:218]: Good catch!

**Comment by ProofOfKeags** [comment]
> I also wonder if settling an invoice before receiving all the htlcs can be problematic from the payer PoV

As far as I can think of, there is only one situation in which this can occur with the current logic. Let's say that the sender pays a total amount `total_msat` to the recipient. In the recipient's original invoice, they set an `amount` in the invoice that they are expecting. This means that with the logic today, the recipient will only begin to settle invoices when the total amount of the received HTLC's exceeds `amount`. In order for any of the payment shards to be swallowed by the routing nodes, without a corresponding outgoing HTLC to the recipient, that payment shard must be smaller than `total_msat - amount`. This is [explicitly prohibited](https://github.com/lightning/bolts/blob/master/04-onion-routing.md?plain=1#L369-L370) in the spec.

Further, even if it was not prohibited by the spec, it's hard to categorize this type of issue as problematic. If the recipient is willing to accept the payment once it exceeds `amount`, they care little about how much the sender actually sent. In the sender's case, the sender was indeed willing to pay the entire amount that they paid out in the MPP. If the sender is happy with what they paid, and the recipient is happy with what they received, I do not believe there is a problem. The "stolen" payment shard can be thought of as large fee to which the sender consented.

**Comment by positiveblue** [comment]
Yes, I am not worried at all about someone else getting the extra funds. I was more interested in the edge case of the sender being able to properly detect that a payment was successful even if one of the htlcs in a MPP errors but the other htlcs in the MPP are enough to settle the invoice.

I think it's an edge case that could happen with really small amounts, and probably we just don't MPP there (4 sat invoice for example)

**Comment by ProofOfKeags** [comment]
> Yes, I am not worried at all about someone else getting the extra funds. I was more interested in the edge case of the sender being able to properly detect that a payment was successful even if one of the htlcs in a MPP errors but the other htlcs in the MPP are enough to settle the invoice.

I see what you are saying. In this case, one of two things will happen. Either the routers will steal a shard, which will appear to the sender as a successful payment, or one of the shards will fail all the way to the source.

In the case where it fails all the way to the source, that shard will either cause the total to dip below the invoice amount or it will not. In the case where it does, presumably an incentive compatible implementation for the recipient will opt to cancel the HTLCs after the timeout period because it does not exceed the threshold (the semantics of which I believe to be preserved by this change) or the recipient will fulfill the HTLC with a preimage reveal, implying that they are satisfied with the result. I believe we can assert that if any of these HTLCs are redeemed (via UpdateFulfillHTLC) then the invoice payment was successful from the POV of the sender.

I think that the edge case you're describing would only present in a scenario where the intermediate hops opted to take a smaller fee and if the size of the payment shards spanned a large enough range. If the max payment shard size is large enough, and the min payment shard size is small enough, then it is possible that the smallest one could be eclipsed by a router opting to take a lower fee on the way. However, since the onion packet specifies `amt_to_forward` at every hop, even if the current fee policy of the router is lower than is implied by the inbound and outbound htlc value difference, it would be irrational for the router to not take the full fee they were offered, *even if they were willing to take a lower one*.

So overall, I don't think there is anything to solve here.

**Comment by morehouse** [comment]
Concept ACK.  We should update/add tests for the new behavior.

File: htlcswitch/link_test.go [hunk 2]
```diff
@@ -807,7 +800,115 @@ func TestExitNodeAmountPayloadMismatch(t *testing.T) {
...
+
+		const amount = btcutil.SatoshiPerBitcoin
+		htlcAmt, htlcExpiry, hops := generateHops(amount,
+			testStartingHeight, n.firstBobChannelLink)
+
+		// In order to exercise this case, we'll now _manually_ modify
+		// the per-hop payload for amount to be the incorrect value.
+		// The acceptable values of the amount to forward should be
+		// lower than the incoming HTLC value.
+		hops[0].FwdInfo.AmountToForward = amount - overpayment
```
Comment (morehouse) [inline htlcswitch/link_test.go:898]: Is there a check somewhere against the original invoice amount?  What happens if both the payload `AmountToForward` and the final hop HTLC amount are below the invoice amount?  Bob *should* fail in that case, but does he?

File: htlcswitch/link_test.go [hunk 3]
```diff
@@ -745,37 +746,28 @@ func TestExitNodeTimelockPayloadMismatch(t *testing.T) {
...
-// TestExitNodeAmountPayloadMismatch tests that when an exit node receives an
-// incoming HTLC, if the amount encoded in the onion payload of the forwarded
-// HTLC doesn't match the expected payment value, then the HTLC will be
-// rejected.
-func TestExitNodeAmountPayloadMismatch(t *testing.T) {
+
+// TestExitNodeTimelockPayloadExceedsHTLC tests that when an exit node receives
+// an incoming HTLC, if the time lock encoded in the payload of the forwarded
+// HTLC isn't lower than the timelock on the incoming HTLC, then the HTLC will
+// be rejected with the appropriate error.
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:769]: You are correct.

File: htlcswitch/link_test.go (hunk 2 above)
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:898]: I believe that is beyond the scope of the changes being made. The logic that I am touching is solely around the final hop onion payload integrity checks. The question I believe you are asking here is how we handle HTLCs that don't cover the full amount of the invoice which is the responsibility of the existing invoice code.

File: htlcswitch/link_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:769]: I will simplify this explanation

File: htlcswitch/link_test.go [hunk 4]
```diff
@@ -745,37 +746,28 @@ func TestExitNodeTimelockPayloadMismatch(t *testing.T) {
 		testStartingHeight, n.firstBobChannelLink)
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:745]: In principle I agree with you but I suddenly am wondering why this was a 3-hop network to begin with, considering I didn't change the overall structure of the test, just the parameters it was measuring. I'd love to understand why this was the way it was.

**Review by saubyk — COMMENTED** [review]
Needs an addition to the release notes (0.17.0) with a short description of the change:
https://github.com/lightningnetwork/lnd/blob/master/docs/release-notes/release-notes-0.17.0.md

**Review by ellemouton — COMMENTED** [review]
Looking good. The other case I think we should think of is AMP payments. With MPP it might be fine for us to consider the set complete once we reach a total amount above `total_msat` since we just release the same pre-image no matter if there was actually another incoming HTLC that we didnt get yet. _But_ with AMP, I think with the current logic of this PR what can happen is: there are 4 HTLCs that the sender sends, we get 3 of them which all are over-paying a bit and together actually make up the total `total_msat` amount. Now the receiver will go ahead and try to reconstruct the pre-images from the 3 shards & will end up with the incorrect pre-images. Yes this is an edge case since the sender isnt allowed to create a shard after already creating a set that meets the total amount _but_ routers nodes may unilaterally increase the HTLC amount. Thoughts?

File: htlcswitch/link_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ellemouton) [reply htlcswitch/link_test.go:745]: i think it is just an available helper?

File: htlcswitch/link_test.go [hunk 5]
```diff
@@ -807,7 +797,113 @@ func TestExitNodeAmountPayloadMismatch(t *testing.T) {
...
+	default:
+		t.Fatalf("incorrect error, expected incorrect cltv expiry, "+
+			"instead have: %v", err)
+	}
+}
+
+// TestExitNodeAmountExceedsHTLC tests that when an exit node receives an
+// incoming HTLC, if the amount encoded in the onion payload of the forwarded
+// HTLC exceeds the incoming payment value, then the HTLC will be rejected.
+func TestExitNodeAmountPayloadExceedsHTLC(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:817]: Can you help me understand why you find it confusing? I find them to be identical ways of saying the same thing.

File: htlcswitch/link_test.go [hunk 6]
```diff
@@ -807,7 +797,113 @@ func TestExitNodeAmountPayloadMismatch(t *testing.T) {
...
+			channels.carolToBob, testStartingHeight)
+		if err := n.start(); err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(n.stop)
+
+		const amount = btcutil.SatoshiPerBitcoin
+		htlcAmt, htlcExpiry, hops := generateHops(amount,
+			testStartingHeight, n.firstBobChannelLink)
+
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:841]: I had to keep them below the line length limit and was following the convention I saw, can you specify what about the formatting you find difficult to read?

**Comment by ProofOfKeags** [comment]
@ellemouton 
>  The other case I think we should think of is AMP payments.

>  Now the receiver will go ahead and try to reconstruct the pre-images from the 3 shards & will end up with the incorrect pre-images

This suggests to me quite strongly that we need to formalize what constitutes a "completed payment". I don't currently know if there is something in the HTLC circuit setup that identifies itself as an AMP. In the discussion I had with @positiveblue above, it appears as though the only way we know if an MPP is completed is whether it exceeds the invoice term threshold. According to your suggestion above this is not the real definition of a completed payment when AMPs are being used. A naive search through this file does not reveal where an AMP preimage is reconstructed, but presumably we already have to make the assessment somewhere that a payment is both an AMP *and* it is complete. Where can that logic be found right now?

File: htlcswitch/link_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ellemouton) [reply htlcswitch/link_test.go:818]: Just cause the title and description is all about "exceeding" but then the actual test is about "underpaying". Yes it is two ways of saying the same thing. Just took me a sec to grok. "exceeding" makes me think that a node is getting more than it expected which is a good thing. So would make more sense to me to say "HLTCUnderpaysPayloadAmount". 

But just my opinion. Non-blocking comment.

**Comment by ProofOfKeags** [comment]
https://github.com/lightningnetwork/lnd/blob/d0ca48a4352892f972a87337274c9193bd81e8ea/invoices/update.go#L267-L273

@ellemouton it looks like the failure block here cancels the HTLCs, which makes sense since if we can't reconstruct the preimages there's not much to do. After reading [this](https://docs.lightning.engineering/lightning-network-tools/lnd/amp) I don't see any obvious way that we can discern whether the AMP was malformed or the last shard hasn't arrived.

However, even if we can't know for sure, in theory what we could do is in the scenario where the current AMP shard sum exceeds the `total_msat` value, but we can't reconstruct the preimage, we could wait up to N {seconds|minutes} for another shard to arrive. If it doesn't then we can do a delayed cancel, which would increase latency for true failures, and in exchange we can forward settle a subset of these scenarios.

All of that said, the machinery to accomplish this is not exactly straightforward and involves adding new asynchronous control flow to the settlement logic, which is a somewhat large expense to the code complexity budget. In addition to that, as you point out above, this is an edge case scenario that requires node operators to act irrationally. Finally, the state of LND prior to this PR doesn't allow overpayment of *any kind* in this scenario, so even without having fixed this edge case, this PR represents a strict improvement over the status quo.

With all of that in mind, I move that we merge this without such logic other than maybe recording how often these AMP reconstruction operations fail. If it turns out that we were wrong about the real customer pain caused by this edge case then we can justify scheduling that work in the future, and if not we can save ourselves the complexity and time.

CC @Roasbeef @saubyk

File: htlcswitch/link_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go]: Fixed

File: htlcswitch/link_test.go [hunk 7]
```diff
@@ -5,21 +5,22 @@ import (
 	"crypto/rand"
 	"crypto/sha256"
 	"encoding/binary"
+	"errors"
```
Comment (yyforyongyu) [inline htlcswitch/link_test.go:8]: git message needs to be `invoices+htlcswitch: add tests....`

File: htlcswitch/link_test.go [hunk 8]
```diff
@@ -721,93 +726,186 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
 	}
 }
 
-// TestExitNodeTimelockPayloadMismatch tests that when an exit node receives an
-// incoming HTLC, if the time lock encoded in the payload of the forwarded HTLC
-// doesn't match the expected payment value, then the HTLC will be rejected
-// with the appropriate error.
-func TestExitNodeTimelockPayloadMismatch(t *testing.T) {
+// TestExitNodeHLTCTimelockExceedsPayload tests that when an exit node receives
```
Comment (yyforyongyu) [inline htlcswitch/link_test.go:729]: 👍

File: htlcswitch/link_test.go [hunk 9]
```diff
@@ -721,93 +726,186 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
...
-	}
+	require.NoError(t, err, "payment should have succeeded but didn't")
 }
 
-// TestExitNodeAmountPayloadMismatch tests that when an exit node receives an
-// incoming HTLC, if the amount encoded in the onion payload of the forwarded
-// HTLC doesn't match the expected payment value, then the HTLC will be
-// rejected.
-func TestExitNodeAmountPayloadMismatch(t *testing.T) {
+// TestExitNodeTimelockPayloadExceedsHTLC tests that when an exit node receives
```
Comment (yyforyongyu) [inline htlcswitch/link_test.go:767]: 👍

File: htlcswitch/link_test.go [hunk 10]
```diff
@@ -721,93 +726,186 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
...
+		_, err = makePayment(
+			n.aliceServer, n.bobServer, firstHop, hops, amount,
+			htlcAmt, htlcExpiry,
+		).Wait(30 * time.Second)
+		assertFailureCode(t, err, lnwire.CodeFinalIncorrectHtlcAmount)
+
+		return err != nil
 	}
-	assertFailureCode(t, err, lnwire.CodeFinalIncorrectHtlcAmount)
+	err := quick.Check(f, &quick.Config{MaxCount: 20})
```
Comment (yyforyongyu) [inline htlcswitch/link_test.go:864]: IMO we should avoid using randomness in our unit tests, plus this randomness doesn't provide more utility here. If we wanna test random inputs we should do it in fuzz testing.

PS, looks like the package `quick` is frozen, so it won't be maintained?

Comment (ProofOfKeags) [reply htlcswitch/link_test.go:864]: I very strongly disagree that testing should not be done with random inputs. Fixed tests do very little to test anything and secondly, by writing tests against random inputs we are forced to test and therefore document the true invariants that are supposed to hold in our code rather than asking the reader to guess at them by trying to do pattern reconstruction.

That said, idk if there's a functional difference here between what I want and what you're calling "fuzz testing" but I generally don't see the value in tests that do not have this setup. There are exceptions of course, but in my experience these are the only types of tests that have helped catch bugs.

Comment (ProofOfKeags) [reply htlcswitch/link_test.go:864]: > Could you elaborate on this part? Coming from TDD background I think it makes little sense.

This will be a long one so bear with me :)

The main purpose of testing is to ensure that your code honors a *contract*. It's purpose is *not* to ensure a particular implementation of that contract. When we design code we expect it to have certain behaviors and it is that behavior we are trying to preserve. The issue with fixed-value unit tests is twofold. First, they only ensure that a singular point in the input space is valid. Secondly, because they allow the test writer to make certain assumptions about things, the test writer will naturally overlook or underdefine the overall contract that the code under test is supposed to honor.

The greatest assurance we can provide that our code is correct would be a proof. However, golang lacks a type system capable of providing proofs of correctness, and even with the languages that afford you this ability, you may still prefer not to write those types because the benefits you get don't outweigh the complexity cost of advanced type system features.

The least assurance we get is from selecting specific test cases: we can only be sure that the case that is tested is correct. There is a wide space of implementations that would satisfy the test case and not the *contract*. But the issue here is that the singular unit test cases are fundamentally incapable of describing a contract because they use a different quantifier.

A unit test case is an assurance claim using a `∃ (there exists)`. i.e. "there exists a value for which this assertion holds". A property test (what you call "fuzz testing") on the other hand attempts to approximate the `∀ (for all)`. i.e. "this assertion holds for all inputs". The claims being tested between these two quantifiers are *fundamentally different*. In fact, they are actually logical negations of one another. It is extremely rare where the genuine contract that your code has claimed to uphold is one where the quantifer is `∃`. In rare cases an individual unit test on a boundary condition (finite integer spaces, zero values, or other "special values" in your algebra) can be very helpful.

Property testing represents a point on this spectrum that gets you most of the way to a proof without having to go through the same rigor you would need to in order to get the same assurances out of a type system. It is for this reason that I believe that property testing should make up 90+% of your test suite. Any other testing strategy gives you comparatively little assurance that anything actually works, and does very little to help you understand how your codebase is *supposed to work*.

So maybe by now you're thinking that all of this is an argument for "just do both". But here's the thing, unit test cases are naturally subsumed by property tests, since the moment you sample your input values from your generation sources, it tests that individual case. So we lose nothing by omitting the unit test case entirely.

Writing code with high assurance, which as I understand from @saubyk is a huge priority to get LND to the point where it is being treated as mission critical software, is about making the strongest assertions you possibly can about the code and unit tests don't offer any benefit over property tests in this regard.

> Back to this specific test case, since we always have hops[0].FwdInfo.AmountToForward = amount + underpayment, how does the randomness help unless you are testing an overflow?

So if we examine this particular case with this perspective, the benefit afforded here is that we are demonstrating that the HTLC is rejected independent of *how much* the Payload is underpaid. This both tests a *stronger assertion* than a unit test ever could, *and* it documents the *actual property* we are looking for out of the code.

Comment (morehouse) [reply htlcswitch/link_test.go:864]: My only concern with randomness in unit tests is reproducibility.  Better to have a consistently failing test in CI than one that might pass on a rerun (and thus hide the bug as "just a flaky test").  @ProofOfKeags Does `quick` seed the PRNG by default?

Besides that, I'm very interested in more property/fuzz testing of LND.

FWIW, this particular usage of `quick` could be trivially rewritten as a Go fuzz test:

```go
func FuzzExitNodeHTLCUnderpaysPayloadAmount(f *testing.F) {
  f.Fuzz(func(underpayment_rand uint64) {
    ...
  })
}
```

Comment (ProofOfKeags) [reply htlcswitch/link_test.go:864]: > Does quick seed the PRNG by default?

Yes. In other language ecosystems I typically see it seeded with the git commit hash. However, I just looked it up and the default implementation of the PRNG seed is to use [the time](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/quick/quick.go;l=198-204)

> My only concern with randomness in unit tests is reproducibility

I agree that repeated runs in the same pipeline should result in the same test failures. However, we actually *do* want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

Comment (morehouse) [reply htlcswitch/link_test.go:864]: > I agree that repeated runs in the same pipeline should result in the same test failures. However, we actually _do_ want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

Our current approach for fuzz tests is to generate an input corpus from many CPU hours of fuzzing and submit them to the `lnd-fuzz` repo.  The LND CI then runs the fuzz tests on their seeds from `lnd-fuzz`.  This gives us determinism for CI and randomness when fuzzing.

One thing this setup is missing is a continuous fuzzing system to keep trying new inputs over time.  Something like https://github.com/lightningnetwork/lnd/issues/5965.

Comment (ProofOfKeags) [reply htlcswitch/link_test.go:864]: Fwiw property testing and fuzz testing [serve different purposes](https://www.tedinski.com/2018/12/11/fuzzing-and-property-testing.html)

I'm not sure what I'm trying to accomplish here is in line with the traditional goals of fuzzing.

I'm trying to introduce (slowly) some algebraic property guarantees for the code that is under test, not necessarily trying to force UB states explicitly. There is significant overlap with these goals but I do want to make clear my aim and why I'm pushing that direction.

Comment (yyforyongyu) [reply htlcswitch/link_test.go:864]: Thanks for the reply and glad we are having this discussion!

> It is for this reason that I believe that property testing should make up 90+% of your test suite.

Ideally yes, in practice no. That'd be a nice goal, but we have limited engineering time. I think the key to making good property testing is to define properties and exhaust them all. Take this specific test case,

1. 20 out of infinite provides very little coverage but too much confidence. We'd think it's protected by property tests while it's not.
2. the state change happens in the test, not in the state machine. In other words, we are testing our test code, not the state machine. This is because we manually change `hops[0].FwdInfo.AmountToForward = amount + underpayment` before testing the state machine, so we are actually testing the addition works as expected, in other words, we are testing a behavior that's not made from the testing object.
3. not every test case needs to be property testing, aside from engineering time, functions like arithmetic operations should not rely on a random behavior but dedicated edge case tests are needed, like zero values, overflow, etc.

Since property test increases the testing time significantly, I don't think it's suitable for every package atm. Maybe we can start with the lower level packages like `lnwallet`.

> Fwiw property testing and fuzz testing serve different purposes

A good read, so is [this one](https://hypothesis.works/articles/what-is-property-based-testing/). Based on the author, the difference comes from where the random data is generated(user or machine).

> Writing code with high assurance

For sure. I think it's nice to have property tests, while I think the more urgent issue is to make our code testable - this means more interfaces, more defined responsibilities for each subsystem, and simply a higher coverage rate.

> However, we actually do want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

I think we need to seed it so it's deterministic. Instead of relying on repeated builds to generate variation, we should increase the input size in the test. Otherwise, here's a simple question - if a developer sees the property test fails in his/her PR, who's gonna fix it? This is already the question we are facing today in our builds, which will soon be fixed.

**Comment by ProofOfKeags** [comment]
> Ideally yes, in practice no. That'd be a nice goal, but we have limited engineering time. 

Right but in this case you're telling me to use *more* engineering time to *remove* this testing practice.

> 20 out of infinite provides very little coverage but too much confidence. We'd think it's protected by property tests while it's not

This argument generalizes to not testing anything at all...I don't think this is a reasonable position to take. I'd love to use the default of 100 but I'd have to be able to reuse the setup and teardown to keep the testing time down. This represents a step in the right direction.

>  in other words, we are testing a behavior that's not made from the testing object.

I think there's a serious misunderstanding about what this test is doing. It is making sure that irrespective of the *amount* that the HTLC overpays the payload amount it will be accepted. This is not simply testing addition. It is testing that the acceptance happens irrespective of the value.

> functions like arithmetic operations should not rely on a random behavior but dedicated edge case tests are needed, like zero values, overflow, etc.

Functions like arithmetic operations benefit the *most* from property testing.. since they have the clearest properties i.e. commutativity, associativity, etc.

> For sure. I think it's nice to have property tests, while I think the more urgent issue is to make our code testable - this means more interfaces, more defined responsibilities for each subsystem, and simply a higher coverage rate.

A culture of property testing *forces* you to test against the interface only since you can't hand pick the test data. Pervasively adopting a property testing mindset is like a more thorough TDD which forces the design of clean and consistent interfaces.

> Otherwise, here's a simple question - if a developer sees the property test fails in his/her PR, who's gonna fix it? This is already the question we are facing today in our builds, which will soon be fixed.

I don't think the fact that we have an underdefined protocol for how to allocate engineering resources is a good argument for making our testing strategy less thorough.

> Since property test increases the testing time significantly

This is the only drawback I can actually see. If this is prohibitive we can turn the number of cases down easily and even if we turned it *all the way down to 1* we still have the benefit of documenting the *real* requirement as opposed to showing an example of the requirement.

File: htlcswitch/link_test.go (hunk 10 above)
Comment (yyforyongyu) [reply htlcswitch/link_test.go:864]: Just to be clear, I'm not saying we should abandon the practice it here, but wanna have more discussion. Fwiw we already have similar tests in [watchtower](https://github.com/lightningnetwork/lnd/blob/master/watchtower/wtdb/codec_test.go#L128), [autopilot](https://github.com/lightningnetwork/lnd/blob/master/autopilot/choice_test.go#L62) and three other packages.

File: htlcswitch/link_test.go [hunk 11]
```diff
@@ -721,93 +726,186 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
...
 		n.aliceServer, n.bobServer, firstHop, hops, amount, htlcAmt,
 		htlcExpiry,
 	).Wait(30 * time.Second)
-	if err == nil {
-		t.Fatalf("payment should have failed but didn't")
+	require.NotNil(t, err, "payment should have failed but didn't")
+
+	var rtErr ClearTextError
+	ok := errors.As(err, &rtErr)
+	require.True(t, ok)
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go]: Fixed

**Review by yyforyongyu — COMMENTED** [review]
Very close! The latest push broke the linter, and the tests failed, otherwise this is LGTM👍

To sum up the discussion, we agree to practice property-based testing, as it provides a more substantial promise of the correctness of the code.

My questions are about the practice of property testing introduced here, writing them down for future reference,

- is the chosen property suitable? Either it's testing 20 or 100 different values, I don't think it's a good property because the sample size is too small. I also don't think we should property test basic arithmetics - if we wanna test an overflow case, we should create such a case instead of relying on a random event from the property tests. This is different from calculating leap years or birthdays, where the dates have their own properties but integers don't(maybe better explain in future PRs).

- should we seed the randomness? In future builds, the current setup only has two outcomes - either it never fails, which means there's little utility, or it fails but, most likely, in other developer's PRs, which creates confusion and is likely to be viewed as a flake and won't be fixed, unless someone is dedicated to monitoring the build.

I'd rather see the tests fail deterministically than succeed randomly, otherwise it'll cause more harm as we'd think we've implemented something correctly while we only get that impression because the test is not robust.

That said, I think we can go with the current approach and see its effect. It's hard to judge without trying them first.

**Comment by ProofOfKeags** [comment]
OK As far as I can tell, all outstanding items have been taken care of on this PR. There are two failing tests still, it doesn't seem like these are related to my changes but it's possible I am missing something.

**Comment by ProofOfKeags** [comment]
I also wanted to ask about the ubuntu itests for etcd and sqlite. I find it difficult to believe the code I changed here broke these tests. Are they known to have issues right now?

**Comment by Roasbeef** [comment]
> With all of that in mind, I move that we merge this without such logic other than maybe recording how often these AMP reconstruction operations fail.

Agreed, this is more or less a bug in the sender not handling their splits or resplits properly. Since it's an AMP payment, they can just try again with distinct values for the `set_id` field and things should eventually work. 

For additional context, this change is intended to enable certain routes to succeed (say due to a `min_htlc`) field where they otherwise wouldn't.

**Comment by Roasbeef** [comment]
> I also wanted to ask about the ubuntu itests for etcd and sqlite. I find it difficult to believe the code I changed here broke these tests. Are they known to have issues right now?

Hmm, yeah those look like existing flakes we should dig in a bit and make issues to track the flake error itself.

**Comment by Roasbeef** [comment]
@yyforyongyu 


> is the chosen property suitable? Either it's testing 20 or 100 different values

So the default value is 100, with the set config: https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/quick/quick.go;l=20.

The core look is actually pretty simple: https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/quick/quick.go;l=281-293
   * Get random values 
   * Try invariant 
   * Stop if max count has been reached


> should we seed the randomness? 

FWIW, we already use `testing/quick` in a few other areas: https://github.com/search?q=repo%3Alightningnetwork%2Flnd+quick.Check&type=code. It's not all that different from other areas where we generate random data, then do some stuff then run an assertion. The difference here is that we can define "generators" for the values we are about, then hook it into the main config, with the assertion being verified up to `N` times. This isn't as advanced as stuff like Haskell's quick check, but can be useful in ensuring an invariant is held over a larger input space than a one time run. 


If the `rand.Rand` value isn't specified, the the package will create one and seed it with the current time: https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/quick/quick.go;l=200-202

File: htlcswitch/link_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/link_test.go]: Yeh was just left around from the past where you wanted to be able to always test the incoming+outgoing logic for a node in the middle.

File: htlcswitch/link_test.go (hunk 2 above)
Comment (Roasbeef) [reply htlcswitch/link_test.go:895]: > Is there a check somewhere against the original invoice amount? What happens if both the payload AmountToForward and the final hop HTLC amount are below the invoice amount? Bob should fail in that case, but does he?

Yes, this is in another sub-system though, found here: https://github.com/lightningnetwork/lnd/blob/86e7b4e1e02e29117dcebba319b0da98e66a3365/invoices/update.go#L196-L200

`inv` is the invoice read from disk. 

The link mainly ensures that the content in the `update_add_htlc` message (created by the incoming node) is congruent with the info in the onion payload (created by the sender).

**Review by saubyk — APPROVED** [review]
Concept Ack

:lgtm:

> *[Reviewable](https://reviewable.io/reviews/lightningnetwork/lnd/7768#-:-NZhqpbHCXXpxMnrrB4t:bdn2tjy)* status: 0 of 5 files reviewed, 19 unresolved discussions (waiting on @ellemouton, @morehouse, @positiveblue, @ProofOfKeags, and @yyforyongyu)

**Review by Roasbeef — APPROVED** [review]
Great PR! I like how you endeavored to improve the testing methodology while working on the change. 

Just a few style nits, and some areas where we can speed up the execution time of tests by not re-creating the entire testing harness with each iteration.

**Comment by Roasbeef** [comment]
Also re the new `fuzz` package vs `testing/quick`, one other big diff is the set of supported types: 
> The fuzzing arguments can only be the following types:
> string, []byte
> int, int8, int16, int32/rune, int64
> uint, uint8/byte, uint16, uint32, uint64
> float32, float64
> bool

While with `testing/quick` the config can be used to define custom generators for arbitrary types. In theory for those types, you can serialize to bytes, then have the test parse that, but that's contextual depending on the nature of the tests (eg: fine for serialization tests, but an add extra burden for say path finding tests).

File: invoices/invoiceregistry_test.go [hunk 12]
```diff
@@ -925,6 +926,84 @@ func TestMppPayment(t *testing.T) {
 	}
 }
 
+// TestMppPaymentWithOverpayment tests settling of an invoice with multiple
+// partial payments. It covers the case where the mpp overpays what is in the
+// invoice.
+func TestMppPaymentWithOverpayment(t *testing.T) {
+	t.Parallel()
+	defer timeout()()
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply invoices/invoiceregistry_test.go:934]: I noticed that there were deferred timeouts in a lot of the adjacent tests so I was trying to follow suit. Good to know that they are not the preferred way of doing things going forward

File: htlcswitch/link_test.go [hunk 13]
```diff
@@ -721,93 +725,184 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
...
+		hops[0].FwdInfo.AmountToForward = amount - overpayment
+		firstHop := n.firstBobChannelLink.ShortChanID()
+		_, err = makePayment(
+			n.aliceServer, n.bobServer, firstHop, hops, amount,
+			htlcAmt, htlcExpiry,
+		).Wait(30 * time.Second)
+
+		return err == nil
+	}
+	err := quick.Check(f, &quick.Config{MaxCount: 20})
```
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/link_test.go:904]: on it

**Review by yyforyongyu — APPROVED** [review]
> I also wanted to ask about the ubuntu itests for etcd and sqlite. I find it difficult to believe the code I changed here broke these tests. Are they known to have issues right now?

Yeah there are two known flakes atm, and we already have PRs to fix them. Soon we'll reach to the point that the builds are green again.
